ktor-server/ktor-server-tests/common/test/io/ktor/tests/server/plugins/CORSTest.kt

```diff
@@ -1275,4 +1291,292 @@ class CORSTest {
+        routing {
+            route("/test") {
+                options {
+                    call.respond("OPTIONS")
+                }
+
+                cors {
+                    allowHost("example.com")
+                }
+
+                get {
+                    call.respond("OK")
+                }
+            }
+        }
+
+        client.options("/test") {
+            header(HttpHeaders.Origin, "https://example.com")
+            header(HttpHeaders.AccessControlRequestMethod, "GET")
+        }.let { response ->
+            assertEquals(response.status, HttpStatusCode.OK)
+            assertEquals(HttpHeaders.Origin, response.headers[HttpHeaders.Vary])
+            assertEquals("https://example.com", response.headers[HttpHeaders.AccessControlAllowOrigin])
+        }
+    }
+
+    @Test
+    fun preflightAllowedMethodsConfinedToRoutes() = testApplication {
+        routing {
+            route("/test") {
+                cors {
+                    allowHost("example.com")
+                    allowMethod(HttpMethod.Put)
+                }
+            }
+
+            route("/other") {
+                cors {
+                    allowHost("example.com")
+                    allowMethod(HttpMethod.Patch)
+                }
+            }
+        }
+
+        client.options("/test") {
+            header(HttpHeaders.Origin, "https://example.com")
+            header(HttpHeaders.AccessControlRequestMethod, "PUT")
+        }.let { response ->
+            assertEquals(response.status, HttpStatusCode.OK)
+            assertEquals(HttpHeaders.Origin, response.headers[HttpHeaders.Vary])
+            assertEquals("https://example.com", response.headers[HttpHeaders.AccessControlAllowOrigin])
+        }
+
+        client.options("/test") {
+            header(HttpHeaders.Origin, "https://example.com")
+            header(HttpHeaders.AccessControlRequestMethod, "PATCH")
+        }.let { response ->
+            assertEquals(response.status, HttpStatusCode.Forbidden)
+        }
+
+        client.options("/other") {
+            header(HttpHeaders.Origin, "https://example.com")
+            header(HttpHeaders.AccessControlRequestMethod, "PATCH")
+        }.let { response ->
+            assertEquals(response.status, HttpStatusCode.OK)
+            assertEquals(HttpHeaders.Origin, response.headers[HttpHeaders.Vary])
+            assertEquals("https://example.com", response.headers[HttpHeaders.AccessControlAllowOrigin])
+        }
+
+        client.options("/other") {
+            header(HttpHeaders.Origin, "https://example.com")
+            header(HttpHeaders.AccessControlRequestMethod, "PUT")
+        }.let { response ->
+            assertEquals(response.status, HttpStatusCode.Forbidden)
+        }
+    }
+
+    @Test
+    fun preflightPluginInParentHandlesChildRoutes() = testApplication {
+        routing {
+            route("/outer") {
+                cors {
+                    anyHost()
```

Comment:
this DSL looks like we're going to write routes inside it:
```
cors {
   route("/abc") {}
}
```

Let's think how to design it right